test/e2e/utils_test.go

```diff
@@ -220,75 +222,109 @@ func getPreviousGitTag(path string, cur semver.Version) (semver.Version, error)
+		return fmt.Sprintf("localhost:%d", kongProxyLocalPort)
+	}
+
+	var extAddrs []string
+	var intAddrs []string
+	nodes, err := env.Cluster().Client().CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	require.NoError(t, err)
+	for _, node := range nodes.Items {
+		for _, naddr := range node.Status.Addresses {
+			if naddr.Type == corev1.NodeExternalIP {
+				extAddrs = append(extAddrs, naddr.Address)
 			}
-		}
-		var extAddrs []string
-		var intAddrs []string
-		nodes, err := env.Cluster().Client().CoreV1().Nodes().List(ctx, metav1.ListOptions{})
-		require.NoError(t, err)
-		for _, node := range nodes.Items {
-			for _, naddr := range node.Status.Addresses {
-				if naddr.Type == corev1.NodeExternalIP {
-					extAddrs = append(extAddrs, naddr.Address)
-				}
-				if naddr.Type == corev1.NodeInternalIP {
-					extAddrs = append(intAddrs, naddr.Address)
-				}
+			if naddr.Type == corev1.NodeInternalIP {
+				intAddrs = append(intAddrs, naddr.Address)
 			}
 		}
-		// local clusters (KIND, minikube) typically provide no external addresses, but their internal addresses are
-		// routeable from their host. We prefer external addresses if they're available, but fall back to internal
-		// in their absence
-		if len(extAddrs) > 0 {
-			proxyIP = fmt.Sprintf("%v:%v", extAddrs[0], port)
-		} else if len(intAddrs) > 0 {
-			proxyIP = fmt.Sprintf("%v:%v", intAddrs[0], port)
-		} else {
-			assert.Fail(t, "both extAddrs and intAddrs are empty")
-		}
 	}
-	return proxyIP
+	// local clusters (KIND, minikube) typically provide no external addresses, but their internal addresses are
+	// routeable from their host. We prefer external addresses if they're available, but fall back to internal
+	// in their absence
+	if len(extAddrs) > 0 {
+		t.Logf("picking an external NodePort address: %s", extAddrs[0])
+		return fmt.Sprintf("%v:%v", extAddrs[0], port.NodePort)
+	} else if len(intAddrs) > 0 {
+		t.Logf("picking an internal NodePort address: %s", intAddrs[0])
+		return fmt.Sprintf("%v:%v", intAddrs[0], port.NodePort)
+	}
+
+	assert.Fail(t, "both extAddrs and intAddrs are empty")
+	return ""
 }
 
-// startPortForwarder runs "kubectl port-forward" in the background. It stops the forward when the provided context
-// ends.
-func startPortForwarder(ctx context.Context, t *testing.T, env environments.Environment, namespace, name, localPort,
-	targetPort string,
-) {
-	kubeconfig, err := generators.NewKubeConfigForRestConfig(env.Name(), env.Cluster().Config())
-	require.NoError(t, err)
-	kubeconfigFile, err := os.CreateTemp(os.TempDir(), "portforward-tests-kubeconfig-")
+// startPortForwarder runs "kubectl port-forward" in the background. It returns a local port that the traffic gets forward to.
+// It stops the forward when the provided context ends.
+func startPortForwarder(ctx context.Context, t *testing.T, env environments.Environment, namespace, name, targetPort string) int {
+	localPort, err := freeport.GetFreePort()
```

Comment:
I like this 👍 

One note though is that this might on very rare occasion fail because between this call and the `kubectl` invocation below someone/something might occupy this port. Hence you might consider (not necessarily change it here, I think we should be good) adding a retry of those 2 calls.

Reply:
It's getting more complicated as the `kubectl port-forward` here is run with `Start` asynchronously so it's not gonna be easy to verify whether the port was occupied without parsing the output somehow. I'll leave it as it is for now, it's still better than hardcoded ports.